File: leetcode/00200-00299/00203_remove-linked-list-elements/Solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func removeElements(head *ListNode, val int) *ListNode {
	if head == nil {
		return head
	}
	head.Next = removeElements(head.Next, val)
	if head.Val == val {
		return head.Next
	}
	return head
}

func removeElements1(head *ListNode, val int) *ListNode {
	dummy := &ListNode{Next: head}

	for tmp := dummy; tmp.Next != nil; {
		if tmp.Next.Val == val {
			tmp.Next = tmp.Next.Next
		} else {
			tmp = tmp.Next
		}
	}
	return dummy.Next
}

func printList(node *ListNode) {
	for node != nil {
		fmt.Println(node.Val)
		node = node.Next
	}
}

func main() {
	var l1, l2, l3, l4 ListNode
	l1.Val = 5
	l2.Val = 1
	l3.Val = 5
	l4.Val = 9
	l1.Next = &l2
	l2.Next = &l3
	l3.Next = &l4
	l4.Next = nil
	head := &l1
	new_head := removeElements(head, 5)
	new_head1 := removeElements1(head, 1)
	printList(new_head)
	printList(new_head1)
}

File: leetcode/00200-00299/00203_remove-linked-list-elements/Solution.ts
/**
 * ID:    00203
 * Title: Remove Linked List Elements
 * Difficulty: Easy
 * Description: Given the head of a linked list and an integer val, remove all the nodes of the linked list that has Node.val == val, and return the new head.
 *
 * Example 1:
 *
 * Input: head = [1,2,6,3,4,5,6], val = 6 Output: [1,2,3,4,5]
 *
 * Example 2:
 *
 * Input: head = [], val = 1 Output: []
 *
 * Example 3:
 *
 * Input: head = [7,7,7,7], val = 7 Output: []
 *
 * Constraints:
 *
 * The number of nodes in the list is in the range [0, 10 4 ].
 * 1 <= Node.val <= 50
 * 0 <= val <= 50
 */
import { ListNode } from "../../utils/ListNode";

function removeElements(head: ListNode | null, val: number): ListNode | null {
  if (!head) return null
  let dummy = new ListNode(0)
  dummy.next = head
  let prev = dummy
  let cur: ListNode | null = head
  while (cur) {
    if (cur.val === val) {
      prev.next = cur.next
    } else {
      prev = cur
    }
    cur = cur.next
  }
  return dummy.next
};

function test_00203() {
  let head = ListNode.create([1, 2, 6, 3, 4, 5, 6]), val = 6
  ListNode.print(removeElements(head, val))
}

test_00203()
